guide_react.rst

```diff
@@ -0,0 +1,550 @@
+ourselves by introducing another functional component called ``Todo``, which 
+will return the text for its corresponding to-do item:
+
+.. code-block:: jsx
+
+    function Todo(props) {
+        const { index, todo } = props;
+        return (
+            <div className="todo">
+            {todo.text}
+            </div>
+        );
+    };
+
+Notice that this functional component has one parameter called ``props``, 
+short for 'properties.' This variable is simply a JavaScript object 
+containing all of the attributes that we pass into the component:
+
+.. code-block:: jsx
+
+    <Todo key={index} index={index} todo={todo} />
+
+From the element above, we can see that there are three variables in props: 
+``key``, ``index``, and ``todo``. However, ``key`` is a special type of prop in 
+React which gives each ``<Todo>`` element in the array a unique identity, 
+so it does not get passed in with the rest of the props. Thus, we can 
+just access the values of ``index`` and ``todo`` when creating our ``Todo`` 
+component by using the following:
+
+.. code-block:: javascript
+
+    const { index, todo } = props
+
+At the end of ``Todo``, we just return a ``<div>`` element containing a 
+string of text from the to-do item.
+
+How to Create a To-Do Item
+**************************
+
+While the code above works for hard coded to-do items, we can improve on this 
+functionality by creating new to-do items with user input. First, we should 
+create a form component that takes in a user’s input and returns it from the 
+``App`` component:
+
+.. code-block:: jsx
+
+    function App() {
+        const [todos, setTodos] = React.useState([]);
+
+        const addTodo = (text) => {
+            const newTodos = [...todos, { text }];
+            setTodos(newTodos);
+        };
+
+        return (
+            <div className="app">
+            <div className="todo-list">
+                {todos.map((todo, index) => (
+                <Todo
+                    key={index}
+                    index={index}
+                    todo={todo}
+                />
+                ))}
+                <TodoForm addTodo={addTodo} />
+            </div>
+            </div>
+        );
+    }
+
+The ``TodoForm`` component only passes in one prop, which is ``addTodo``—a 
+function that will add a new to-do item to the `todos` array. We can define 
+``TodoForm`` like so:
+
+.. code-block:: jsx
+
+    const TodoForm = (props) => {
+        const { addTodo } = props;
+        const [value, setValue] = React.useState("");
+
+        const handleSubmit = (e) => {
+            e.preventDefault();
+            if (!value) return;
+            addTodo(value);
+            setValue("");
+        };
+
+        return (
+            <form onSubmit={handleSubmit}>
+            <input
+                type="text"
+                className="input"
+                value={value}
+                onChange={(e) => setValue(e.target.value)}
```

Comment:
what does this do? some more explanation in the following paragraph might be helpful